modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+
+        if (failure == null) {
+            assert opCtx0 != null : ctx.localNodeId();
+
+            Map<Integer, StoredCacheData> globalCfgs = new HashMap<>();
+
+            for (List<StoredCacheData> storedCfgs : res.values()) {
+                if (storedCfgs == null)
+                    continue;
+
+                for (StoredCacheData cacheData : storedCfgs)
+                    globalCfgs.put(CU.cacheId(cacheData.config().getName()), cacheData);
+            }
+
+            opCtx0.cfgs = globalCfgs;
+
+            if (U.isLocalNodeCoordinator(ctx.discovery()))
+                cacheStartProc.start(reqId, reqId);
+
+            return;
+        }
+
+        if (opCtx0 == null)
+            finishProcess(failure);
+        else // Remove files asynchronously.
+            ctx.cache().context().snapshotMgr().snapshotExecutorService().execute(() -> {
+                rollback(opCtx0);
+
+                finishProcess(failure);
+            });
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @return Result future.
+     */
+    private IgniteInternalFuture<Boolean> cacheStart(UUID reqId) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 == null)
+            return new GridFinishedFuture<>();
+
+        if (!reqId.equals(opCtx0.reqId)) {
+            return new GridFinishedFuture<>(
+                new IgniteCheckedException("Unknown snapshot restore operation was rejected."));
+        }
+
+        if (!U.isLocalNodeCoordinator(ctx.discovery()))
+            return new GridFinishedFuture<>();
+
+        DiscoveryDataClusterState state = ctx.state().clusterState();
+
+        if (state.state() != ClusterState.ACTIVE || state.transition())
+            return new GridFinishedFuture<>(new IgniteCheckedException(OP_REJECT_MSG + "The cluster should be active."));
+
+        Throwable err = opCtx0.err.get();
+
+        if (err != null)
+            return new GridFinishedFuture<>(err);
+
+        if (!allNodesInBaselineAndAlive(opCtx0.nodes))
+            return new GridFinishedFuture<>(new IgniteCheckedException(OP_REJECT_MSG + "Server node(s) has left the cluster."));
+
+        GridFutureAdapter<Boolean> retFut = new GridFutureAdapter<>();
+
+        try {
+            Collection<StoredCacheData> ccfgs = opCtx0.cfgs.values();
+
+            // Ensure that shared cache groups has no conflicts before start caches.
+            for (StoredCacheData cfg : ccfgs) {
+                if (!F.isEmpty(cfg.config().getGroupName()))
+                    ensureCacheAbsent(cfg.config().getName());
+            }
+
+            if (log.isInfoEnabled()) {
+                log.info("Starting restored caches " +
+                    "[requestID=" + opCtx0.reqId + ", snapshot=" + opCtx0.snpName +
+                    ", caches=" + F.viewReadOnly(ccfgs, c -> c.config().getName()) + ']');
+            }
+
+            ctx.cache().dynamicStartCachesByStoredConf(ccfgs, true, true, false, null, true, opCtx0.nodes).listen(
```

Comment:
Can you return the `IgniteInternalFuture` directly here or `chain` the results if you need any additional result transformation?

Reply:
Done